README.md
# **NES MMC5 Demo**

A demo to test and show features that could be integrated in a NES game using the MMC5 chip.

![MMC5_Demo_a_v3](MMC5_Demo_a_v3_2.gif)

----------------

## Compile

`make`

----------------

## Run

`make run_a`


**Important**: Because this demo use more than 32K of PRG RAM and none of the ExROM boards was made with this amount of RAM,
**Some emulator may not be able to run the demo**. The game will stop and blink red if it cannot access more than 32K of RAM.

----------------

### FCEUX emulator

FCEUX can natively emulate a MMC5 with more than 32K of RAM and will run the demo without problems.

FCEUX download link: <https://fceux.com/web/download.html>

----------------

### Mesen emulator

If you want to use Mesen for his debugging capability. You can still use it by "making some modification". For this, you will need to:

1. Download the source of Mesen-X (fork of Mesen and most up-to-date version that I am aware of): <https://github.com/NovaSquirrel/Mesen-X>
2. Follow the instruction on how to compile Mesen: <https://github.com/NovaSquirrel/Mesen-X/blob/master/COMPILING.md>
3. Change some lines of code in the file named "MMC5.h" located in the "Core" folder:

    - Change line 80 from:

        ```c++
        bankNumber &= 0x07;
        ```

        to:

        ```c++
        bankNumber &= 0x0F;
        ```

    - Change around line 100 from:

        ```c++
        if(realSaveRamSize + realWorkRamSize != 0x4000 && bankNumber >= 4) {
            //When not 2x 8kb (=16kb), banks 4/5/6/7 select the empty socket and return open bus
            accessType = MemoryAccessType::NoAccess;
        }
        ```

        to:

        ```c++
        // if(realSaveRamSize + realWorkRamSize != 0x4000 && bankNumber >= 4) {
        //  //When not 2x 8kb (=16kb), banks 4/5/6/7 select the empty socket and return open bus
        //  accessType = MemoryAccessType::NoAccess;
        // }
        ```

4. Recompile Mesen and if everything was done correctly, the demo should not blink red anymore and start normally.


CHANGELOG.md
# CHANGELOG

The format of this CHANGELOG is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

This project try to follow the [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

-----------------

## **[0.4.0]** - _2022-08-11_

### **Added**

#### Global

- A GIF of the demo in the README.
- Sprites draw buffer.
- 3 CHR ROM of 256K:
  - empty: just zeroes.
  - text: with a short compressed text.
  - ui: contain a font and 2 tiles for the dialog box borders.
- Some constants.
- Some variables (many about the dialog box).
- A sign in CHR ROM.

#### Data

- Huffman code for each text characters.
- A little dictionary and the pointers for the 2 example dialogs.
- A single sign object.
- 2 local dialog entity.
- 2 sign with their entity in the level.

#### Module

- A dialog module which:
  - Split the screen for x scanlines to the other VRAM screen and draw the (still compressed) text.
  - Functions to be able to read and process raw text data (load_raw_data, next_bit, next_char).
  - Add a function for each stage of the dialog box (init, clear, border, main, next, end)
- Local entities can now be loaded and called.
- An 'ai' for the dialog entity (check if the player collide and run the dialog)
- Some function in the world module (scrBufIdx2entityBufAdr, scrBufIdx2int)

#### Python

- empty_char.py: Generate an empty CHR ROM.
- encode.py: Encode a text file into binary using dictionary compression. Return dictionary as txt and asm.
- huffman_encode.py: Encode the dictionary compressed text into a Huffman form. Return char code and dialog pointer as txt and asm.
- decode.py: decode the Huffman compressed text and print the result.
- gen_lorem.py: generate a lorem ipsum text file.
- A bat file to run the 2 encoder scripts.

#### Vector

- NMI flag to read from PPU.
- The variable "nmi_mmc5_nametable" to set nametables before any NMI actions.

### **Changed**

#### Global

- Entity buffers.
- Refactor entity buffer.
- Sprite variables.
- CHR size to the max of 1024K (1M).
- Gitignore for python generated files.
- Some constants.

#### Data

- Player object data.
- Screen object to have the random mask.
- Last background palette in the level.
- Lower chance of random tile in the screen object.

#### Module

- Reorganize the code into more files.
- Move entity action into a separate module.
- Modules priority (tile_anim + 1, load_screens - 1)
- Refactor:
  - Screen fill objects loading mainly to not tell animating empty tiles
    and be faster for fill with no random.
  - How entities are loaded and drawn.
  - The "draw_player_move" function.
  - Screen draw function to also clear the second VRAM screen.
  - Sprites to be 8\*8 instead of 8\*16 (take less space in CHR)

#### Vector

- Refactor scanline IRQs to handle UI separation.
- NMI flags order.
- Reset to set attributes to last palette.

### **Removed**

- C part in Makefile.
- Useless variables (scrbuf_update_array_idx, scr_index).
- The macro res_entity_buf

### **Fixed**

- (Maybe ?) objects not being drawn at the correct location or having some wrong tiles.
- Reset not clearing the last RAM bank.
- Game state not waiting for the load level module to finished.
- The taskTime script drawing the graph too high when scanline was above 240.
- Input bug when moving.
- Loading objects when switching banks between entity and background.

-----------------

## **[0.3.0]** - _2022-07-21_

The _Player movements and collisions_ update

### **Added**

#### Global

- A Lua script to show the current loaded screens and another to show modules working time.
- An infinite loop that make the screen blink red if the game cannot access more than 32K of PRG RAM.
- Entity buffers and 3 other variables (sprite_banks, global_entity_spr_counter and entity_load_counter).
- Some constants (RAM_MAX_BNK, PRGRAM_SPR_BANK, SCANLINE_BOT, SCANLINE_TOP and SCANLINE_MID).
- A macro (res_entity_buf).
- A basic player texture in CHR.

#### Data

- A player entity that can move and control the scroll.
- 1 player entity in the first screen of the level.

#### Module

- A function to load global entity with a position.
- In control module:
  - 4 functions "check_scroll_XXX" to check if the player can move in each direction.
    It is also checking if the front tile has a collision.
  - Functions to get and set the player position.
  - A "draw_player_move" function to update the animation of the player.
  - A "setScroll2PlayerPos" subroutine to help with collisions.
- A new module "draw_global_sprites" that draw global entities on the screen.

#### Utils

- New operations: dec_tmp, sub_tmp and sub_16 (16 bits sub).

### **Changed**

#### Global

- Makefile and MMC5_Demo_mmc5.cfg file.
- Update [README](README.md).
- PRG RAM to 64K (_if the game blink red, check the [README](README.md)_).
- Maximum number of lower module to 8 and maximum priority to 4.
- Change BTN_TIMER to 8.
- Enable sprites.
- In CHR ROM:
  - Water tiles.
  - All tiles position to fit the collision system.

#### Data

- The level sprite palettes and add sprite banks data to fit the new level structure.
- All object tiles to fit the collision system.

#### Module

- Modules priority.
- Object pos-size code to another file.
- Draw object function to also "draw" (load) entities.
- Move "scroll2scrBufAdr" out of the world module.
- Control module by:
  - Making the input subroutine run by the player entity and not the control module anymore.
  - Ticking global entity and run the "draw global entity" module.
  - Refactoring "input" function to "player_input".
  - Changing scroll functions to take into account player position and update its position and sprites.
- In loading modules:
  - The loading screen module to load entities and reset the screen entity buffers.
  - The loading level to load the level sprite banks and reset global entity buffer.

#### Vector

- Scanline IRQs to:
  - Use the new constants.
  - Hide the first row of tiles instead of the last one.

asm/constant.asm
;**********
; Constants
;**********

;----------
; PPU
;----------
PPU_CTRL   := $2000
PPU_MASK   := $2001
PPU_STATUS := $2002
PPU_SCROLL := $2005
PPU_ADDR   := $2006
PPU_DATA   := $2007

; PPU MASK
PPU_MASK_GREY = %00000001
PPU_MASK_BKG8 = %00000010
PPU_MASK_SPR8 = %00000100
PPU_MASK_BKG  = %00001000
PPU_MASK_SPR  = %00010000
PPU_MASK_R    = %00100000
PPU_MASK_G    = %01000000
PPU_MASK_B    = %10000000

; PPU CTRL
PPU_CTRL_NM_1     = %00000001
PPU_CTRL_NM_2     = %00000010
PPU_CTRL_INC      = %00000100
PPU_CTRL_SPR      = %00001000
PPU_CTRL_BKG      = %00010000
PPU_CTRL_SPR_SIZE = %00100000
PPU_CTRL_SEL      = %01000000
PPU_CTRL_NMI      = %10000000



;----------
; APU
;----------
APU := $4000

APU_SQ1_VOL   := $4000
APU_SQ1_SWEEP := $4001
APU_SQ1_LO    := $4002
APU_SQ1_HI    := $4003

APU_SQ2_VOL   := $4004
APU_SQ2_SWEEP := $4005
APU_SQ2_LO    := $4006
APU_SQ2_HI    := $4007

APU_TRI_LINEAR := $4008
APU_TRI_LO     := $400A
APU_TRI_HI     := $400B

APU_NOISE_VOL := $400C
APU_NOISE_LO  := $400E
APU_NOISE_HI  := $400F

APU_DMC_FREQ  := $4010
APU_DMC_RAW   := $4011
APU_DMC_START := $4012
APU_DMC_LEN   := $4013

APU_SND_CHN := $4015
APU_CTRL    := $4015
APU_STATUS  := $4015
APU_FRAME   := $4017



;----------
; OAM
;----------
OAMDMA := $4014



;----------
; IO
;----------
IO_JOY1 := $4016
IO_JOY2 := $4017


;----------
; NMI
;----------
NMI_DONE = %10000000
NMI_MAIN = %01000000
NMI_SCRL = %00100000
NMI_READ = %00010000
NMI_PLT  = %00001000
NMI_ATR  = %00000100
NMI_SPR  = %00000010
NMI_BKG  = %00000001


;----------
; MMC5
;----------
.ifdef MMC5
; PRG Mode
MMC5_PRG_MODE  := $5100
; CHR Mode
MMC5_CHR_MODE  := $5101
; RAM protection
MMC5_RAM_PRO1  := $5102
MMC5_RAM_PRO2  := $5103
; Extended RAM mode
MMC5_EXT_RAM   := $5104
; Nametable mapping
MMC5_NAMETABLE := $5105
; Fill nametable
MMC5_FILL_TILE := $5106
MMC5_FILL_COL  := $5107
; RAM Bank
MMC5_RAM_BNK   := $5113
; PRG Banks
MMC5_PRG_BNK0  := $5114
MMC5_PRG_BNK1  := $5115
MMC5_PRG_BNK2  := $5116
MMC5_PRG_BNK3  := $5117
; CHR Banks
MMC5_CHR_BNK0  := $5120
MMC5_CHR_BNK1  := $5121
MMC5_CHR_BNK2  := $5122
MMC5_CHR_BNK3  := $5123
MMC5_CHR_BNK4  := $5124
MMC5_CHR_BNK5  := $5125
MMC5_CHR_BNK6  := $5126
MMC5_CHR_BNK7  := $5127
MMC5_CHR_BNK8  := $5128
MMC5_CHR_BNK9  := $5129
MMC5_CHR_BNKA  := $512A
MMC5_CHR_BNKB  := $512B
MMC5_CHR_UPPER := $5130
; Vertical Split
MMC5_SPLT_MODE := $5200
MMC5_SPLT_SCRL := $5201
MMC5_SPLT_BNK  := $5202
; IRQ Scanline counter
MMC5_SCNL_VAL  := $5203
MMC5_SCNL_STAT := $5204
; Unsigned 8x8 to 16 Multiplier
MMC5_MUL_A     := $5205
MMC5_MUL_B     := $5206
; Expansion RAM
MMC5_EXP_RAM   := $5C00
; RAM
MMC5_RAM       := $6000
.endif


;----------
; Game
;----------
.define HIGHER_MODULE_SIZE  2
.define LOWER_MODULE_SIZE  8
.define LOWER_MODULE_MAX_PRIO  4

.define MODULE_CTRL  0+$80
.define MODULE_WORLD 1+$80
.define MODULE_DIALOG 0+$80

.define WORLD_BANK_START 2

.define NB_PRGRAM_BNK 1

.define OBJ_BANK 4+$80

.enum STATE
    START
    WAIT
    DRAW_SCREEN
    DRAW_SCREEN_WAIT
    LOAD_SCR_ALL
    NORMAL
    DIALOG
.endenum

.define BTN_TIMER 8

.enum DIR
    UP
    DOWN
    LEFT
    RIGHT
.endenum

.enum INPUT
    _A      = %10000000
    _B      = %01000000
    SELECT  = %00100000
    START   = %00010000
    UP      = %00001000
    DOWN    = %00000100
    LEFT    = %00000010
    RIGHT   = %00000001
.endenum

.define SCANLINE_MID $40
.define SCANLINE_BOT $41
.define SCANLINE_TOP $02
.define SCANLINE_UI  $43
.define SCANLINE_UI_END $44

.define DIALOG_BOX_HEIGHT 56

.define PRGRAM_LEVEL_BANK 0
.define PRGRAM_SCREEN_BANK 1
.define PRGRAM_ENTITY_BANK 4
.define PRGRAM_SPR_BANK 4

.define SCR_UPD_SCR_X_R $A0
.define SCR_UPD_SCR_X_L $61
.define SCR_UPD_SCR_Y_D $90
.define SCR_UPD_SCR_Y_U $70

.define NAMETABLE_SCROLL %00000000
.define NAMETABLE_FILL   %11111111
.define NAMETABLE_ALL    %11100100
.define NAMETABLE_UI     %01010101

.define ANIM_BASE_SPD_MASK %00001111
.define ANIM_MAX_FRAME_MASK %00000011

.define RAM_MAX_BNK 7

.define PPU_RD_BUF_SIZE 16

.define ANIM_NONE $00
.define ANIM_FLIP $40
.define ANIM_INC  $80

.enum CHAR
    _0
    _1
    _2
    _3
    _4
    _5
    _6
    _7
    _8
    _9
    _A
    _B
    _C
    _D
    _E
    _F
    _G
    _H
    _I
    _J
    _K
    _L
    _M
    _N
    _O
    _P
    _Q
    _R
    _S
    _T
    _U
    _V
    _W
    _X
    _Y
    _Z
    SPACE
    COMMA
    EXCLAMATION
    INTEROGATION
    RES_1 = $38
    RES_2
    NL
    SPD
    DICT
    LDICT
    EXT
    END
.endenum
